File: matlab/gridSearch2.m
function [optimalAllocation parSet] = gridSearch2(cumPnl, netPosSystem, netPosAcct, af, qt, pnlAcctSoFar, pricePre, priceNow, onePtValue, scale, criteriaOption, sanityOnNetPosition)

sgnNet = sign(netPosSystem);
par  = af*abs(netPosSystem);
par  = round(par);

% sanity check
% there are many different ways of doing this
% floor all and then add to those with biggest remainder
% until you get the original sum back
% at the end the method is indifferent
while (sum(par) ~= abs(netPosSystem))
    if sum(par) > abs(netPosSystem)
        index = find(par == max(par));
        if length(index)>1
            index = find(par./af == max(par./af));            
        end
        par(index(1)) = par(index(1))-1;
        %par(index(end)) = par(index(end))-1;
    else
        index = find(par == max(par));
        if length(index)>1
            index = find(par./af == max(par./af));            
        end
        par(index(1)) = par(index(1))+1;
        %par(index(end)) = par(index(end))+1;
    %disp(index)
    end
end

netPosAcct2 = sgnNet*par;
qtAcct = netPosAcct2 - netPosAcct;
parSet.optimalAllocation = abs(qtAcct);

% fprintf("Net position ");
% for j=1:6
%     fprintf('%d ',netPosAcct(j));
% end
% 
% fprintf(" Before: ");
% for j=1:6
%     fprintf('%d ',netPosAcct2(j));
% end
% fprintf('\n');

netPosAcct2 = sanityCheckOnNetPosAcct(netPosAcct, netPosAcct2, scale, qt);
% 
% fprintf('After: ');
% for j=1:6
%     fprintf('%d ',netPosAcct2(j));
% end
% fprintf('\n');



qtAcct = netPosAcct2 - netPosAcct;
parSet.optimalAllocation = abs(qtAcct);

if (scale < 0 && sum(qtAcct>0)>0) || (scale > 0 && sum(qtAcct<0)>0) || sum(qtAcct) ~= scale*qt
    disp('ERROR: In gridSreach2 - something is not right ...');
end

parSet.netPosAcct2 = netPosAcct2;
parSet.sgnNet = sgnNet;
parSet.af = af;
parSet.cumPnl = cumPnl;
parSet.pnlAcctSoFar = pnlAcctSoFar;
parSet.netPosAcct = netPosAcct;
parSet.scale = scale;
parSet.qt = qt;
parSet.pricePre = pricePre;
parSet.priceNow = priceNow;
parSet.onePtValue = onePtValue;
parSet.criteriaOption = criteriaOption;
parSet.sanityOnNetPosition = sanityOnNetPosition;
       
pnlAcctSoFar2 = pnlAcctSoFar + netPosAcct2*(priceNow-pricePre)*onePtValue;
parSet.minMae = criterion(pnlAcctSoFar2, cumPnl, af, '1');

parSet.optimalCase = [];
parSet.netSoFar = netPosAcct2;

parSet.perturbVec = [+1 -1];
parSet = optimizingPnlByPerturbingNetPos(parSet);

parSet.perturbVec = [+1 -1 +1 -1];
parSet = optimizingPnlByPerturbingNetPos(parSet);

parSet.perturbVec = [-1 -1 +2];
parSet = optimizingPnlByPerturbingNetPos(parSet);

parSet.perturbVec = [+1 +1 -2];
parSet = optimizingPnlByPerturbingNetPos(parSet);

parSet.perturbVec = [+2 -2];
parSet = optimizingPnlByPerturbingNetPos(parSet);

parSet.perturbVec = [+3 -3];
parSet = optimizingPnlByPerturbingNetPos(parSet);

parSet.perturbVec = [+1 +2 -3];
parSet = optimizingPnlByPerturbingNetPos(parSet);

parSet.perturbVec =	[+3 -1 -2];
parSet = optimizingPnlByPerturbingNetPos(parSet);

optimalAllocation = parSet.optimalAllocation;
%

File: bak/gridSearch2.cpp
/*
* @Author: shayaan
* @Date:   2020-06-11 23:15:02
* @Last Modified by:   shayaan
* @Last Modified time: 2020-06-16 00:28:21
*/

#include "allInclude.h"

static parSet gridSearch2(double cumPnl, double netPosSystem, vector <double> netPosAcct, vector <double> af, int qt,
                   vector <double> pnlAcctSoFar, double pricePre, double priceNow, int onePtValue, int scale,
                   char criteriaOption, double sanityOnNetPosition){

    parSet parSet(af.size(),af,onePtValue);

    int sgnNet;
    if ( netPosSystem > 0){
        sgnNet = 1; 
    }
    if ( netPosSystem < 0){
        sgnNet = -1; 
    }
    
    vector <double> par;
    par.clear();
    
    for (int i = 0; i < af.size(); i++){
        par.push_back(round (af[i] * fabs( netPosSystem )));
    }

    while ( accumulate(begin(par), end(par), 0) != fabs(netPosSystem)){

        if (accumulate(begin(par), end(par), 0) > fabs(netPosSystem)){
            int max = par[0]; //setting seed for maximum
            vector<int> maxIndex = {0};
            for(int i = 0; i < par.size(); i++){
                if (par[i] > max){
                    max = par[i];
                    maxIndex.clear();
                    maxIndex.push_back(i);
                }
                if (par[i] == max){
                    maxIndex.push_back(i);
                }
            }

            int INDEX;
            if(maxIndex.size() > 1){
                int minaf = 1;

                for(int i = 0; i < maxIndex.size(); i++){
                    if(af[maxIndex[i]] < minaf){
                        minaf = af[maxIndex[i]];
                        INDEX = i;
                    }
                }
            } 

            else{
                INDEX = maxIndex[0];
            }
            par[INDEX] -= 1;

        }

        else{
            int max = par[0];
            vector<int> maxIndex = {0};
            for(int i = 0; i < par.size(); i++){
                if (par[i] > max){
                    max = par[i];
                    maxIndex.clear();
                    maxIndex.push_back(i);
                }
                if (par[i] == max){
                    maxIndex.push_back(i);
                }
            }
            int INDEX;
            if(maxIndex.size() > 1){
                int minaf = 1;

                for(int i = 0; i < maxIndex.size(); i++){
                    if(af[maxIndex[i]] < minaf){
                        minaf = af[maxIndex[i]];
                        INDEX = i;
                    }
                }
            } else{
                INDEX = maxIndex[0];
            }
            par[INDEX] += 1;

        }
    }


    vector<double> netPosAcct2;
    netPosAcct2.clear();
    for (int i = 0; i < par.size(); i++){
        netPosAcct2.push_back(sgnNet*par[i]);
    }

    vector<double> qtAcct;
    qtAcct.clear();


    for(int i = 0; i < netPosAcct2.size(); i++){

        qtAcct.push_back(netPosAcct2[i] - netPosAcct[i]);

    }



    for(int i = 0; i < qtAcct.size(); i++){
        parSet.optimalAllocation[i] = fabs(qtAcct[i]);
    }

    netPosAcct2 = sanityCheckOnNetPosAcct(netPosAcct, netPosAcct2, scale, qt);
    
    qtAcct.clear();

    for(int i = 0; i < netPosAcct2.size(); i++){
        qtAcct.push_back(netPosAcct2[i] - netPosAcct[i]);
    }

    for(int i = 0; i < qtAcct.size(); i++){
        parSet.optimalAllocation[i] = fabs(qtAcct[i]);
    }

    double sumqtAcct = 0;

    for(int i = 0; i < qtAcct.size(); i++){
        sumqtAcct += qtAcct[i];
    }

    if((scale < 0 && sumqtAcct > 0)>0 || (scale > 0 && sumqtAcct < 0) > 0 || sumqtAcct != scale * qt){
        cout << "ERROR: In gridSreach2 - something is not right ..." << endl;
    }


    parSet.netPosAcct2 = netPosAcct2;
    parSet.sgnNet = sgnNet;
    parSet.af = af;
    parSet.cumPnl = cumPnl;
    parSet.pnlAcctSoFar = pnlAcctSoFar;
    parSet.netPosAcct = netPosAcct;
    parSet.scale = scale;
    parSet.qt = qt;
    parSet.pricePre = pricePre;
    parSet.priceNow = priceNow;
    parSet.onePtValue = onePtValue;
    parSet.criteriaOption = criteriaOption;
    parSet.sanityOnNetPosition = sanityOnNetPosition;

    vector <double> pnlAcctSoFar2;
    pnlAcctSoFar2.clear();

    for (int i = 0; i < pnlAcctSoFar.size(); i++ ){
        pnlAcctSoFar2.push_back( pnlAcctSoFar[i] + netPosAcct2[i] * (priceNow - pricePre) * onePtValue); // PnL with new positions

    }



    parSet.minMae = criterion(pnlAcctSoFar2, cumPnl, af, criteriaOption-'0');




    parSet.optimalCase.clear();
    parSet.netSoFar = netPosAcct2;


    parSet.perturbVec = {1, -1};
    parSet.optimizingPnlByPerturbingNetPos();
    // cout<<"One Done"<<endl;

    parSet.perturbVec = {1, -1, 1, -1};
    parSet.optimizingPnlByPerturbingNetPos();
    // cout<<"Two Done"<<endl;
    parSet.perturbVec = {-1, -1, 2};
    parSet.optimizingPnlByPerturbingNetPos();

    parSet.perturbVec = {1, 1, -2};
    parSet.optimizingPnlByPerturbingNetPos();

    parSet.perturbVec = {2,-2};
    parSet.optimizingPnlByPerturbingNetPos();

    parSet.perturbVec = {3, -3};
    parSet.optimizingPnlByPerturbingNetPos();

    parSet.perturbVec = {1, 2, -3};
    parSet.optimizingPnlByPerturbingNetPos();

    parSet.perturbVec = {3, -1, -2};
    parSet.optimizingPnlByPerturbingNetPos();

    return parSet;

}
